File: Heap-Sort/python/leetcode_lc_215.py
nums = [3,2,1,5,6,4]
k = 2

class Solution:
    def findKthLargest(self, nums, k: int) -> int:
        for i in range(int(len(nums)/2)-1,-1,-1):
            self.heapify(nums,i,len(nums))

        for i in range(len(nums)-1,-1,-1):
            k -= 1
            if k == 0 :
                return nums[0]

            nums[0],nums[i] = nums[i],nums[0]
            self.heapify(nums,0,i)

        return nums[0]

    def heapify(self,arr,idx,n):
        left = 2*idx + 1
        right = 2*idx + 2
        parent = idx

        if left < n and arr[left] > arr[parent]:
            parent = left

        if right < n and arr[right] > arr[parent]:
            parent = right

        if parent != idx:
            arr[idx],arr[parent] = arr[parent],arr[idx]
            self.heapify(arr,parent,n)



solution = Solution()

print("数组{} 第 {} 个最大的元素为: {}".format(nums,k,solution.findKthLargest(nums,k)))

File: Heap-Sort/js/leetcode_lc_215.js
const nums = [3,2,1,5,6,4]
const k = 2

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function(nums, k) {
    for(let i = Math.floor(nums.length/2)-1 ; i >=0 ; i--){
        heapify(nums,i,nums.length)
    }

    for(let i = nums.length-1 ; i >= 0 ; i--){
        k -= 1
        if(k == 0){
            return nums[0]
        }
        const tmp = nums[i]
        nums[i] = nums[0]
        nums[0] = tmp
        heapify(nums,0,i)
    }
    return nums[0]
};


var heapify = function(arr,idx,len){
    const left = 2*idx + 1
    const right = 2*idx + 2
    let parent = idx

    if(left < len && arr[left] > arr[parent]){
        parent = left
    }

    if(right < len && arr[right] > arr[parent]){
        parent = right
    }

    if(parent !== idx){
        const tmp = arr[idx]
        arr[idx] = arr[parent]
        arr[parent] = tmp
        heapify(arr,parent,len)
    }
}


console.log("预期是5,结果为:",findKthLargest(nums,k))  // 预期是5